internal/topology/variables/cluster_variable_defaulting_test.go

```diff
@@ -0,0 +1,317 @@
+package variables
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+func Test_DefaultClusterVariables(t *testing.T) {
+	tests := []struct {
+		name                  string
+		clusterClassVariables []clusterv1.ClusterClassVariable
+		clusterVariables      []clusterv1.ClusterVariable
+		want                  []clusterv1.ClusterVariable
+		wantErr               bool
+	}{
+		{
+			name: "Default one variable of each valid type",
+			clusterClassVariables: []clusterv1.ClusterClassVariable{
+				{
+					Name:     "cpu",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "integer",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`1`)},
+						},
+					},
+				},
+				{
+					Name:     "location",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "string",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`"us-east"`)},
+						},
+					},
+				},
+
+				{
+					Name:     "count",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "number",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`0.1`)},
+						},
+					},
+				},
+				{
+					Name:     "correct",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "boolean",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`true`)},
+						},
+					},
+				},
+			},
+			clusterVariables: []clusterv1.ClusterVariable{},
+			want: []clusterv1.ClusterVariable{
+				{
+					Name: "cpu",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`1`),
+					},
+				},
+				{
+					Name: "location",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"us-east"`),
+					},
+				},
+				{
+					Name: "count",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`0.1`),
+					},
+				},
+				{
+					Name: "correct",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`true`),
+					},
+				},
+			},
+		},
+		{
+			name: "Don't default variables that are set",
+			clusterClassVariables: []clusterv1.ClusterClassVariable{
+				{
+					Name:     "cpu",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "integer",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`1`)},
+						},
+					},
+				},
+				{
+					Name:     "correct",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "boolean",
+							Default: &apiextensionsv1.JSON{Raw: []byte(`true`)},
+						},
+					},
+				},
+			},
+			clusterVariables: []clusterv1.ClusterVariable{
+				{
+					Name: "correct",
+
```

Comment:
```suggestion
```
nit

Reply:
Going to leave this as it emphasises the comment here.